src/provider/form-factory-provider.tsx

```diff
@@ -0,0 +1,138 @@
+import { validateForm } from './form-factory-helper';
+
+interface FormFactoryProviderContextProps {
+  patient: fhir.Patient;
+  sessionMode: SessionMode;
+  sessionDate: Date;
+  formJson: FormSchema;
+  formProcessors: Record<string, FormProcessorConstructor>;
+  layoutType: LayoutType;
+  workspaceLayout: 'minimized' | 'maximized';
+  visit: OpenmrsResource;
+  location: OpenmrsResource;
+  provider: OpenmrsResource;
+  registerForm: (formId: string, context: FormContextProps, isSubForm: boolean) => void;
+  getSubForms: () => FormContextProps[];
+  getRootForm: () => FormContextProps;
+  setCurrentPage: (page: string) => void;
+  handleConfirmQuestionDeletion?: (question: Readonly<FormField>) => Promise<void>;
+}
+
+interface FormFactoryProviderProps {
+  patient: fhir.Patient;
+  sessionMode: SessionMode;
+  sessionDate: Date;
+  formJson: FormSchema;
+  workspaceLayout: 'minimized' | 'maximized';
+  location: OpenmrsResource;
+  provider: OpenmrsResource;
+  visit: OpenmrsResource;
+  children: React.ReactNode;
+  formSubmissionProps: {
+    isSubmitting: boolean;
+    setIsSubmitting: (isSubmitting: boolean) => void;
+    onSubmit: (data: any) => void;
+    onError: (error: any) => void;
+    handleClose: () => void;
+  };
+  setCurrentPage: (page: string) => void;
+  handleConfirmQuestionDeletion?: (question: Readonly<FormField>) => Promise<void>;
+}
+
+const FormFactoryProviderContext = createContext<FormFactoryProviderContextProps | undefined>(undefined);
+
+export const FormFactoryProvider: React.FC<FormFactoryProviderProps> = ({
+  patient,
+  sessionMode,
+  sessionDate,
+  formJson,
+  workspaceLayout,
+  location,
+  provider,
+  visit,
+  children,
+  formSubmissionProps,
+  setCurrentPage,
+  handleConfirmQuestionDeletion,
+}) => {
+  const rootForm = useRef<FormContextProps>();
+  const subForms = useRef<Record<string, FormContextProps>>({});
+  const layoutType = useLayoutType();
+  const { isSubmitting, setIsSubmitting, onSubmit, onError, handleClose } = formSubmissionProps;
+  const registerForm = (formId: string, context: FormContextProps, isSubForm: boolean) => {
+    if (isSubForm) {
+      subForms.current[formId] = context;
+    } else {
+      rootForm.current = context;
+    }
+  };
+  // TODO: Manage and load processors from the registry
+  const formProcessors = useRef<Record<string, FormProcessorConstructor>>({
+    EncounterFormProcessor: EncounterFormProcessor,
+  });
+  const getSubForms = () => Object.values(subForms.current);
+  const getRootForm = () => rootForm.current;
+
+  useEffect(() => {
+    if (isSubmitting) {
+      // TODO: find a dynamic way of managing the forms processing order
+      const forms = [rootForm.current, ...getSubForms()];
+      // validate all forms
+      const valid = forms.every((formContext) => validateForm(formContext));
+      if (valid) {
+        Promise.all(
+          forms.map((formContext) => formContext.processor.processSubmission(formContext, new AbortController())),
```

Comment:
I always wonder about AbortControllers like this. If there's no point at which the controller can be cancelled, its not really worth defining one.

Reply:
Just updated the component to fire an abort signal when the component unmounts.